### DISCUSSION TITLE
Constraints

### DISCUSSION THREAD

**Discussion by imwaymaran** [MIT-Emerging-Talent/ET6-foundations-group-05#7] [General]
Constraints discussion

**Comment by misik-eng** [comment]
I'll write example constraints in here, and we can choose or create new ones , after we are all done one of us can create a branch and commit this constraints.md markdown and create a pull request. then we can review it and merge it.

# Constraints

## External Constraints
- Let's set a deadline for the project and aim to complete it at least 4 days before the final submission date to allow time for reviews.
- The programming language is fixed as Python.
- Coding, documentation, and reviews will adhere to established rules based on reviewed examples.
- Pushing to branches is allowed, but pushing directly to the main branch is prohibited. Even if you try, it won't work since this rule is enforced in the repository. The only way to make changes to the main branch is by creating a pull request.
- Each pull request must have at least one review; otherwise, it cannot be merged (this rule is enforced in the repository).
- Everyone will find their own challenges from the internet and add them to a project board list. For now, each person should find two simple (or medium-level, so we can all handle them) Python challenges. Examples can be found on platforms like LeetCode or Codewars.
- Since there are 5 team members, each member will complete at least 2 challenges and review 2 challenges. However, because there are only 5 people, some already-reviewed challenges might need to be reviewed again. This is fine, and additional feedback or comments can be added during the second review.

## Internal Constraints: Involuntary
- Team members have varying levels of knowledge about Git and GitHub, so we should ask each other for help whenever we encounter issues. To facilitate this, we can create issues in the "Issues" section. A practical example of this can be demonstrated during or after the next meeting.
- Not all members may always be available, but it's beneficial for everyone to review and attempt to solve questions as soon as they can, even if they don't know the answer immediately.
- Some members (like me) may have limited experience in coding and code reviews.
- The final deadline for the project is January 10, 2025.

## Internal Constraints: Voluntary
- Naming conventions for branches are important to avoid confusion or unclear naming.
  **How to prevent issues:**
  - Avoid meaningless or confusing names. For example, number the challenges to ensure clarity; if everyone writes "challenge," it will cause confusion.
- The team will use an asynchronous working model and hold regular meetings to track progress.

**Comment by misik-eng** [comment]
Send here your change or new constraint and we will be ready on wednesday to add it to our repo

**Comment by doctorbanu** [comment]
Banu

1. Everyone has different schedule. So, sometimes it’s hard to find regular meeting times. But we will catch up on anything we miss and all contribute equally.(Internal - involuntary)

2. Some of us, myself included, are still learning in certain areas, so we will likely need to lean on each other more and collaborate more as we learn. (Internal - involuntary)

**Comment by GeehanAli** [comment]
**External Constraints: **
Time Pressure: Limited time to study a large amount of new information, terms, websites, and materials.

**Internal Constraints: Involuntary: **
Team Availability: Team members' availability is impacted by other ongoing commitments.

**Internal Constraints: Voluntary: **
Scope Limitation: Deliberately restricting the project scope to keep it manageable and prevent unnecessary complexity.

**Comment by imwaymaran** [comment]
# Constraints

Some boundaries around our project 

## External Constraints:
- **Deadline:** We have a strict deadline to meet.
- **Time Pressure:** Limited time to study a large amount of new information, terms, 
  websites, and materials.
- **Fixed Programming Language:** The programming language for this project is Python 
  and cannot be changed.
- **Coding Standards:** All coding, documentation, and reviews will adhere to 
  established rules based on reviewed examples.
- **Restricted Pushes:** We’re not allowed to push directly to the `master` branch.
- **Branch Merging Rules:** A branch can only be merged after a pull request has been 
  reviewed and approved.
- **Team Size:** There are only 5 people on our team. This can be an advantage but 
  also a limitation if we need more help.

## Internal Constraints (Involuntary):
- **Different Schedules:** Everyone has their own commitments, so it’s hard to find 
  regular meeting times. But we’ll make sure to catch up and contribute equally.
- **Learning Together:** Some of us are still building our skills, so we’ll need to 
  collaborate and support each other more as we learn.
- **Limited Availability:** Balancing this project with other responsibilities might 
  sometimes make it hard to stay fully available.
- **Knowledge Gaps:** Some team members may lack experience in specific technologies 
  or concepts being used in the project, requiring extra time for learning or 
  mentoring.
- **Dependency on Others:** Progress might depend on tasks or decisions made by other 
  team members or external stakeholders.
- **Limited Review Capacity:** Fewer people available to review pull requests or 
  contributions, potentially slowing down approvals and progress.

## Internal Constraints (Voluntary):
- **Keeping It Simple:** We’re deliberately keeping the project scope small to make it 
  easier to manage and avoid unnecessary complexity.
- **Following Rules:** We’ve decided to follow clear coding, documentation, and 
  review rules to maintain quality and consistency.
- **Team Collaboration:** We’re committed to working together and ensuring everyone 
  has a chance to contribute equally, even if it takes more time.
